Add dnssnoop plugin

lorddoskias opened the pull request:
Now that the tcpsnoop is in place and alongside it the build framework for bpf program here's the dnssnoop one.

lorddoskias committed:
Add dnsnoop plugin

It works by creating a raw socket and attaching an ebpf program to
filter only incoming dns packets. That way replies can be parsed.

Signed-off-by: Nikolay Borisov <nborisov@suse.com>

lorddoskias committed:
Use BTF_PATH as the env variable when providing external btf files

Firstyear reviewed (commented):
I'm not really seeing how the go code talks to the bpf bit to register the hooks, but I think that's something I can ask you about later so I can learn more :) 

Otherwise looks really straight forward!

jeffmahoney commented on vql/linux/dnssnoop/dnssnoop.bpf.c:
```diff
@@ -0,0 +1,66 @@
+	if (skb->pkt_type != PACKET_HOST)
+		return 0;
+
+	if (bpf_skb_load_bytes(skb, offsetof(struct ethhdr, h_proto), &protocol, 2) < 0)
```
Does this mean it ends up being ethernet-specific? That'll cover most cases, but could end up leaving e.g. VPNs uncovered.

lorddoskias replied:
I think this indeed could miss i.e l3 interfaces (like vpns). But i think the problem stems from the fact the way raw sockets are implemented. Looking at the documentation for AF_PACKET: 

```
When protocol is set to htons(ETH_P_ALL), then all
       protocols are received.  All incoming packets of that protocol
       type will be passed to the packet socket before they are passed
       to the protocols implemented in the kernel.
```

jeffmahoney commented on vql/linux/dnssnoop/dnssnoop.go:
```diff
@@ -0,0 +1,165 @@
+				log.Fatalf("Error reading from socket\n")
+			}
+
+			packet := gopacket.NewPacket(received_data, layers.LayerTypeEthernet, gopacket.Default)
```
Again, ethernet-specific. Would it make sense to peel back the L2 header in the bpf code so that we get the IP-level packet here?

lorddoskias replied:
According to gopacket's documentation:

`Packets can be decoded from a number of starting points. Many of our base types implement Decoder, which allow us to decode packets for which we don't have full data.`

So in this case I just start decoding at the ethernet layer. Nothing prevents the gocode to be turned into: 

```
// Decode a TCP header and its payload
tcpP := gopacket.NewPacket(p3, layers.LayerTypeTCP, gopacket.Default)
```

For more info on gopacket check:  [Basic Usage](https://pkg.go.dev/github.com/google/gopacket#hdr-Basic_Usage) 

So this might actually be better, however it doesn't solve the problem that the raw socket might be indeed receiving l2 packets.

lorddoskias committed:
dnssnoop: Don't try to close module when creating it failed

This results in a runtime segmentation fault as bpfModule would be null
at this point.

lorddoskias committed:
dnssnoop: Fix handling of packets on l3 interfaces

Handle pure l3 interfaces by trying to parse ipv4/6 packets in case we
fail to get a valid dns packet by trying to parse from l2 header.

lorddoskias committed:
Introduce bpflib

As the number of ebpf-based helpers is growing let's factor out common
code to ease maintenance and reduce duplication.

lorddoskias committed:
dnssnoop: Switch to using newly introduced bpflib

lorddoskias committed:
tcpsnoop: Switch to using newly introduced bpflib

jeffmahoney reviewed (approved):
LGTM
